classes/ActionScheduler_wpPostStore.php

```diff
@@ -462,25 +465,101 @@ protected function generate_claim_id() {
 	}
 
 	/**
-	 * @param string $claim_id
-	 * @param int $limit
+	 * @param string   $claim_id
+	 * @param int      $limit
 	 * @param DateTime $before_date Should use UTC timezone.
+	 * @param string   $group       Claim only actions in the given group.
+	 *
 	 * @return int The number of actions that were claimed
 	 * @throws RuntimeException
 	 */
-	protected function claim_actions( $claim_id, $limit, DateTime $before_date = NULL ) {
+	protected function claim_actions( $claim_id, $limit, DateTime $before_date = null, $group = '' ) {
 		/** @var wpdb $wpdb */
 		global $wpdb;
 
-		$date = is_null($before_date) ? as_get_datetime_object() : clone $before_date;
-		// can't use $wpdb->update() because of the <= condition, using post_modified to take advantage of indexes
-		$sql = "UPDATE {$wpdb->posts} SET post_password = %s, post_modified_gmt = %s, post_modified = %s WHERE post_type = %s AND post_status = %s AND post_password = '' AND post_date_gmt <= %s ORDER BY menu_order ASC, post_date_gmt ASC, ID ASC LIMIT %d";
-		$sql = $wpdb->prepare( $sql, array( $claim_id, current_time('mysql', true), current_time('mysql'), self::POST_TYPE, 'pending', $date->format('Y-m-d H:i:s'), $limit ) );
-		$rows_affected = $wpdb->query($sql);
+		$date = null === $before_date ? as_get_datetime_object() : clone $before_date;
+
+		// Set up a query for post IDs to use later.
+		$query      = new WP_Query();
+		$query_args = array(
+			'fields'           => 'ids',
+			'post_type'        => self::POST_TYPE,
+			'post_status'      => ActionScheduler_Store::STATUS_PENDING,
+			'has_password'     => false,
+			'posts_per_page'   => $limit,
+			'suppress_filters' => true,
+			'no_found_rows'    => true,
+			'orderby'          => array(
+				'menu_order' => 'ASC',
+				'date'       => 'ASC',
+				'ID'         => 'ASC',
+			),
+			'date_query'       => array(
+				'column' => 'post_date',
+				array(
+					'compare' => '<=',
+					'year'    => $date->format( 'Y' ),
+					'month'   => $date->format( 'n' ),
+					'day'     => $date->format( 'j' ),
+					'hour'    => $date->format( 'G' ),
+					'minute'  => $date->format( 'i' ),
+					'second'  => $date->format( 's' ),
+				),
+			),
+		);
+
+		// Check for a particular group.
+		if ( ! empty( $group ) && term_exists( $group, self::GROUP_TAXONOMY ) ) {
+			$query_args['tax_query'] = array(
+				array(
+					'taxonomy'         => self::GROUP_TAXONOMY,
+					'field'            => 'slug',
+					'terms'            => $group,
+					'include_children' => false,
+				),
+			);
+		}
+
+		// Gather relevant IDs.
+		$ids = $query->query( $query_args );
```

Comment:
There is an overheard in running this query. Especially on large databases. Because of that, I'd only use it when we need to filter by group and do a `'tax_query'` rather than every time a claim is staked, regardless of whether this query is needed or not.

Reply:
I understand. That would necessitate dividing up this function to either use `WP_Query` with the `tax_query` and `date_query`, or doing it the old way. I'll take a look at architecting that to avoid duplication as much as possible.